src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -229,54 +230,76 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+
+  private def deleteWorkspace(workspaceName: WorkspaceName, workspaceContext: SlickWorkspaceContext): Future[PerRequestMessage] = {
     //Attempt to abort any running workflows so they don't write any more to the bucket.
     //Notice that we're kicking off Futures to do the aborts concurrently, but we never collect their results!
     //This is because there's nothing we can do if Cromwell fails, so we might as well move on and let the
     //ExecutionContext run the futures whenever
-    dataAccess.workflowQuery.findWorkflowsByWorkspace(workspaceContext).result.map { recs => recs.map {
-      case wf if !WorkflowStatuses.withName(wf.status).isDone && wf.externalId.isDefined => executionServiceCluster.abort(wf, userInfo).map {
-        case Failure(regrets) =>
-          logger.info(s"failure aborting workflow ${wf.externalId} while deleting workspace ${workspaceName}", regrets)
-          Failure(regrets)
-        case success => success
-      }
-    }} andThen {
-      DBIO.from(gcsDAO.deleteBucket(workspaceContext.workspace.bucketName, bucketDeletionMonitor))
-    } andThen {
-      val groupRefs: Set[RawlsGroupRef] = workspaceContext.workspace.accessLevels.values.toSet ++ workspaceContext.workspace.realmACLs.values
-      DBIO.seq(groupRefs.map { groupRef => dataAccess.rawlsGroupQuery.load(groupRef) flatMap {
-        case Some(group) => DBIO.from(gcsDAO.deleteGoogleGroup(group))
-        case None => DBIO.successful(Unit)
-      }}.toSeq: _*)
-    } andThen {
-      DBIO.seq(dataAccess.workspaceQuery.deleteWorkspaceAccessReferences(workspaceContext.workspaceId))
-    } andThen {
-      DBIO.seq(dataAccess.workspaceQuery.deleteWorkspaceSubmissions(workspaceContext.workspaceId))
-    } andThen {
-      DBIO.seq(dataAccess.workspaceQuery.deleteWorkspaceMethodConfigs(workspaceContext.workspaceId))
-    } andThen {
-      dataAccess.workspaceQuery.deleteWorkspaceEntitiesAndAttributes(workspaceContext.workspaceId)
-    } andThen {
-      DBIO.seq(workspaceContext.workspace.accessLevels.map { case (_, group) =>
-        dataAccess.rawlsGroupQuery.delete(group)
-      }.toSeq: _*)
-    } andThen {
-      dataAccess.workspaceQuery.delete(workspaceName)
-    } andThen {
-      DBIO.successful(RequestComplete(StatusCodes.Accepted, s"Your Google bucket ${workspaceContext.workspace.bucketName} will be deleted within 24h."))
+    val x: Future[(Seq[WorkflowRecord], String, Seq[Option[RawlsGroup]])] = dataSource.inTransaction { dataAccess => {
+      for {
+        workflowsToAbort <- dataAccess.workflowQuery.findActiveWorkflowsWithExternalIds(workspaceContext)
+        _ <- dataAccess.workflowQuery.findWorkflowsByWorkspace(workspaceContext).result.map { recs => recs.collect {
+          case wf if !WorkflowStatuses.withName(wf.status).isDone && wf.externalId.isDefined => executionServiceCluster.abort(wf, userInfo).map {
```

Comment:
a for-comp with a `db.query.result map { stuff => more things` in it ... this is getting difficult to read. can you break some of it out into an internal function?